### PR TITLE
Coerce data to text for JSON parsing

### PR DESCRIPTION
**Pull request by jakirkham**
Cleans up some Python 2/3 code for handling JSON parsing by simply always coercing metadata to text regardless of Python version.

xref: https://github.com/zarr-developers/zarr/pull/372
xref: https://github.com/zarr-developers/zarr/pull/401

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [x] Docs build locally (e.g., run ``tox -e docs``)
* [x] AppVeyor and Travis CI passes
* [x] Test coverage is 100% (Coveralls passes)

### CODE REVIEW

File: zarr/storage.py [hunk 1]
```diff
@@ -2296,15 +2296,7 @@ def __getitem__(self, key):
...
-            value = doc[self._value]
-
-            # Coerce `bson.Binary` to `bytes` type on Python 2.
-            # PyMongo handles this conversion for us on Python 3.
-            # ref: http://api.mongodb.com/python/current/python3.html#id3
-            if PY2:  # pragma: py3 no cover
-                value = binary_type(value)
-
-            return value
+            return doc[self._value]
```
Comment (jhamman) [inline zarr/storage.py]: this makes me very happy to see

Comment (jakirkham) [reply zarr/storage.py]: Likewise. 🙂

FWIW it turns out this is not Python 2 specific. We just only handled decoding before parsing JSON on Python 3 (hence avoiding the issue there). With this change we just always decode to text before parsing JSON. Here's a short reproducer.

```python
>>> import json
>>> json.loads(b"{}")
{}
>>> json.loads(b"{\x00}")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jkirkham/miniconda/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/Users/jkirkham/miniconda/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/Users/jkirkham/miniconda/lib/python3.7/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)
```

**Comment by jakirkham** [comment]
Planning on merging end of day Friday if no comments.

**Comment by jakirkham** [comment]
It’s unrelated. Though I agree it’s important to fix. Let’s discuss after.

**Comment by jakirkham** [comment]
Thanks all 😄

**Pull request merged**